### PR TITLE
Edge image types for RHEL 8.5

### PR DESCRIPTION
**Pull request by achilleas-k**
This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [x] adequate documentation informing people about the change such as
  - [x] create a file in [news/unreleased](https://github.com/osbuild/osbuild-composer/tree/main/docs/news/unreleased) directory if this change should be mentioned in the release news
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

## Summary

Adding new distro RHEL 8.5 with support for the three Edge image types: `edge-commit`, `edge-container`, and `edge-installer`.

## New pipeline definition

The definition of the image types differs from older implementations:
- Reusable functions define alternatives to the `pipelines()` function for different image types.  These are bound to the image type instance on creation.  This pattern avoids the frequent conditional branching we ended up having in older pipeline generation functions.  This is especially visible in the definition of the `edge-installer` pipelines that share almost no common components with the other types (only the build pipeline is the same).  Similarly, the other (general) image types will be quite different from the edge image types when they're added, so they can be defined in their own functions while reusing common pipelines.
- Pipelines, options, and inputs are in different files in the package.  A minor change, but I think this makes navigating the code easier.

We could have each image type have its own struct implementation of the `ImageType` interface with its own `pipelines()` method, but then we would need to rewrite all the other common methods as well (e.g., `Arch()`, `Name()`, `PackageSets()`).  I prefer the way I did it now, but I'm open to hearing different opinions on this.

I'm quite happy with the way the pipelines are defined now.  It makes it very easy to look at a `pipelines()` function and see the list of pipelines that it contains.  Same goes for looking at every individual pipeline and seeing the stages.  

A big part of the redesign came out of discussions with @thozza (thanks Tomáš) about having a more declarative implementation of the manifests in osbuild-compser.

## Image names

As discussed in other channels, the new image types are named `edge-*`, without the `rhel-` prefix that they have in 8.4.

## Fixes to 8.4

There's a small fix for the `rhel-edge-container` image type in 8.4: The commit's parent is no longer added to the manifest's sources.

### CODE REVIEW

**Comment by jkozol** [comment]
Great stuff! The naming changes will require an update in cockpit-composer before it can release into 8.5. I captured this [here](https://github.com/osbuild/cockpit-composer/issues/1283) and will soon create a PR.

**Comment by ondrejbudai** [comment]
Looks good so far! :)

**Comment by achilleas-k** [comment]
Update on the status here:
- Testing 8.5 needs an osbuild nightly build.
- osbuild nightly build depends on https://github.com/osbuild/osbuild/pull/635 being merged so we can pin the post-merge commit here.
- osbuild PR is failing on openstack boot tests.

**Comment by ondrejbudai** [comment]
osbuild/osbuild#635 merged :)

File: internal/distroregistry/distroregistry.go
Comment (teg) [inline internal/distroregistry/distroregistry.go:24]: Centos Should be moved to 8.5 now too, right?

File: internal/distro/rhel85/distro.go
Comment (teg) [inline internal/distro/rhel85/distro.go:168]: Don't we also include the kernel here, or do I misremember?

File: internal/distro/rhel85/distro.go
Comment (teg) [inline internal/distro/rhel85/distro.go:442]: Bug fix of previous commit?

File: internal/distro/rhel85/distro.go
Comment (teg) [inline internal/distro/rhel85/distro.go:498]: Ditto?

File: internal/distro/rhel84/distro_v2.go
Comment (teg) [inline internal/distro/rhel84/distro_v2.go:130]: Even though changing old distro definitions should be a no-no, and also should have no effect as we don't plan to do any more 8.4 releases, I think this is ok, as it only avoids a download but otherwise should have no effect. Want to point out that in general we should avoid it though.

**Review by teg — COMMENTED** [review]
Looks great! I just had a few questions, no need to rebase unless there are actual issues though.

**Comment by achilleas-k** [comment]
Thanks @ondrejbudai.

Pinned osbuild to the post-merge commit and changed the condition in mockbuild.sh to match your comments on the corresponding change in osbuild.

File: internal/distroregistry/distroregistry.go
Comment (achilleas-k) [reply internal/distroregistry/distroregistry.go:24]: Yup, that's the plan.  I think we should do it in a separate PR though.  I wanted to keep this one short to see what people think of the way I wrote the image type definitions and pipelines.

File: internal/distro/rhel85/distro.go
Comment (achilleas-k) [reply internal/distro/rhel85/distro.go:168]: The kernel comes with the blueprint packages: https://github.com/osbuild/osbuild-composer/blob/cf73edd2d2248bbb5e561a8beeab694857be2132/internal/blueprint/blueprint.go#L107
including the fallback when unspecified:
https://github.com/osbuild/osbuild-composer/blob/cf73edd2d2248bbb5e561a8beeab694857be2132/internal/blueprint/customizations.go#L165-L167

Comment (teg) [reply internal/distro/rhel85/distro.go:168]: Ah, of course. I should know that ;)

File: internal/distroregistry/distroregistry.go
Comment (teg) [reply internal/distroregistry/distroregistry.go:24]: Makes sense. I guess we also need to implement all the other image types for 8.5 first too.

File: internal/distro/rhel85/distro.go
Comment (achilleas-k) [reply internal/distro/rhel85/distro.go:442]: Yes.  Rebased to keep the history clean.

File: internal/distro/rhel85/distro.go
Comment (achilleas-k) [reply internal/distro/rhel85/distro.go:498]: Ditto

**Review by thozza — COMMENTED** [review]
I like how you structured the code and split it into separate files. I think that this version is good for 8.5, however I think that we should take this approach further.

In order to make addition of new non-edge image types simpler, I would propose to somehow transform `ostreeTreePipeline()` function into a generic function, which can be used by any image type. My reasoning is that it creates stages for all potential Blueprint customizations and Image options. Therefore this will be the most commonly used function by all image types.

Please update `tools/test-case-generators/distro-arch-imagetype-map.json` with new distribution and proper image types.

Otherwise this looks very good to me. Thanks for all your effort!

File: internal/distro/rhel85/distro.go
Comment (achilleas-k) [reply internal/distro/rhel85/distro.go]: Good point.  Changed and rebased.

**Comment by achilleas-k** [comment]
> In order to make addition of new non-edge image types simpler, I would propose to somehow transform `ostreeTreePipeline()` function into a generic function, which can be used by any image type. My reasoning is that it creates stages for all potential Blueprint customizations and Image options. Therefore this will be the most commonly used function by all image types.

I've already started with this kind of renaming and reuse for a followup PR.  Most of the `ostreeTreePipeline()` function is going to become a set of common image stages (I'm going with the name `coreStages` for now), but some parts, the preptree stage in particular, are ostree specific.

I can include the changes in this PR, but the purpose of the separation wont become clear until I include the tar as well.  That said, I could add the tar-installer here as well, but it'll change the scope of the PR.

**Comment by thozza** [comment]
> > In order to make addition of new non-edge image types simpler, I would propose to somehow transform `ostreeTreePipeline()` function into a generic function, which can be used by any image type. My reasoning is that it creates stages for all potential Blueprint customizations and Image options. Therefore this will be the most commonly used function by all image types.
> 
> I've already started with this kind of renaming and reuse for a followup PR. Most of the `ostreeTreePipeline()` function is going to become a set of common image stages (I'm going with the name `coreStages` for now), but some parts, the preptree stage in particular, are ostree specific.

Great, thus sounds exactly like what I was thinking about.

> 
> I can include the changes in this PR, but the purpose of the separation wont become clear until I include the tar as well. That said, I could add the tar-installer here as well, but it'll change the scope of the PR.

Follow-up PR is fine by me. I just think that we should do it rather sooner than later, mainly because we should start producing official 8.5 KVM guest images as well as eventually the AWS EC2 images for 8.5.

Thanks!

**Comment by achilleas-k** [comment]
Removed the explicit enabling of greenboot services (rebased).

See [rhbz#1935177](https://bugzilla.redhat.com/show_bug.cgi?id=1935177)

**Comment by achilleas-k** [comment]
> Removed the explicit enabling of greenboot and redboot services (rebased).
> 
> See [rhbz#1935177](https://bugzilla.redhat.com/show_bug.cgi?id=1935177)

ostree tests are failing because of this change.  The services should now be enabled by the package rhel-release, but it hasn't been updated to include the fix yet.

File: internal/distro/rhel85/distro.go
Comment (ondrejbudai) [inline internal/distro/rhel85/distro.go:16]: I think we decided to change the name to `rhel-8.5`, right? Should we do it straight away?

File: test/data/repositories/rhel-85.json
Comment (ondrejbudai) [inline test/data/repositories/rhel-85.json:5]: Why don't we use rpmrepo?

**Review by ondrejbudai — COMMENTED** [review]
Looks great, two comments inline.

File: test/data/repositories/rhel-85.json
Comment (thozza) [reply test/data/repositories/rhel-85.json:5]: Just a note, that we actually never used it for unreleased RHEL versions. We do the same for 8.4...

File: internal/distro/rhel85/distro.go
Comment (thozza) [reply internal/distro/rhel85/distro.go:16]: Just a note, that this will affect Weldr and other parts, which rely on the name to be without a dot. Something to consider, but I agree that it would make sense to start using the dot.

Comment (achilleas-k) [reply internal/distro/rhel85/distro.go:16]: I started the work with the new naming scheme on the branch and realised that the distro name detection would need to change, or we would need a condition special for 8.5.  Since the concept of aliases came up in that discussion (which is a good idea), I think we should do it in a separate PR where we add support for distro aliases and rename or alias existing distros.

File: test/data/repositories/rhel-85.json
Comment (achilleas-k) [reply test/data/repositories/rhel-85.json:5]: Like @thozza said.  I went with what we used for 8.4.

Comment (ondrejbudai) [reply test/data/repositories/rhel-85.json:5]: Ah, I guess it makes sense. Let's change it if it makes us some troubles.

File: internal/distro/rhel85/distro.go
Comment (ondrejbudai) [reply internal/distro/rhel85/distro.go:16]: Makes sense!

**Review by ondrejbudai — APPROVED** [review]
Awesome!

File: test/cases/ostree.sh
Comment (ondrejbudai) [inline test/cases/ostree.sh:41]: This needs to `8.5`, schutzbot is failing on it. :( Can you please fix it?

**Review by ondrejbudai — CHANGES_REQUESTED** [review]

File: test/cases/ostree.sh
Comment (achilleas-k) [reply test/cases/ostree.sh]: Thank you.  Fixed.

**Comment by yih-redhat** [comment]
@achilleas-k  Do you know when will composer-cli use the new names 'edge-commit, edge-container, edge-installer'?

**Comment by achilleas-k** [comment]
> @achilleas-k  Do you know when will composer-cli use the new names 'edge-commit, edge-container, edge-installer'?
composer-cli reads the image names from the API so there's no need to build in support.  If you run this branch on 8.5 and call `composer-cli compose types` it will list the supported image types from composer.

File: test/cases/ostree.sh
Comment (atodorov) [inline test/cases/ostree.sh:41]: `BOOT_LOCATION="$COPOSE_URL/compose/BaseOS/x86_64/os/"`

otherwise we're not properly testing against specific composes, e.g. rel-eng snapshots. See #1388

**Review by atodorov — CHANGES_REQUESTED** [review]

**Comment by henrywang** [comment]
@achilleas-k according to [errata](https://errata.devel.redhat.com/advisory/68370#c14) and [redhat-release-8.5-0.3.el8.x86_64.rpm](http://download-node-02.eng.bos.redhat.com/rhel-8/nightly/RHEL-8/latest-RHEL-8.5.0/compose/BaseOS/x86_64/os/Packages/redhat-release-8.5-0.3.el8.x86_64.rpm), we can re-run test and land this PR if [`bz#1935177`](https://bugzilla.redhat.com/show_bug.cgi?id=1935177) is the only blocker.

File: test/cases/ostree.sh
Comment (atodorov) [inline test/cases/ostree.sh:36]: FYI https://github.com/osbuild/osbuild-composer/pull/1388#issuecomment-833330668

**Comment by achilleas-k** [comment]
Currently working on splitting out the foundational parts of this PR.  I will open a separate PR that includes the 8.5 distro support without any of the image types.  This is to unblock other image types that have been waiting for this one to merge.

**Comment by achilleas-k** [comment]
Rebased on main.

File: test/cases/ostree.sh
Comment (atodorov) [inline test/cases/ostree.sh:40]: This is in conflict/at odds with https://github.com/osbuild/osbuild-composer/issues/1397 and also partially with https://github.com/osbuild/osbuild-composer/pull/1274. Need to be careful not to end up with locations pointing both to 8.4 and 8.5 nightlies.

File: tools/provision.sh
Comment (atodorov) [inline tools/provision.sh:64]: Related to https://github.com/osbuild/osbuild-composer/issues/1318

File: test/data/repositories/rhel-85.json
Comment (achilleas-k) [reply test/data/repositories/rhel-85.json]: Thanks.  Fixed :)

File: test/cases/ostree.sh
Comment (atodorov) [reply test/cases/ostree.sh:40]: Please rebase again and drop line 40.

Comment (achilleas-k) [reply test/cases/ostree.sh]: Done

**Comment by henrywang** [comment]
`error: Sanity-checking final rpmdb: Didn't find package 'greenboot-failing-unit-1.0-1.el8.noarch'` error in RHEL 8.5 `ostree-ng.sh` test is caused by [rpm-ostree issue](https://bugzilla.redhat.com/show_bug.cgi?id=1959874)

**Comment by achilleas-k** [comment]
Rebased on `main` and cleaned up commits relating to the old Jenkins CI.  Need to define a RHEL 8.5 runner in https://github.com/osbuild/gitlab-ci-terraform first.

**Comment by achilleas-k** [comment]
Added 8.5 to GitLab CI, but https://github.com/osbuild/gitlab-ci-terraform/pull/9 should be merged first

File: .gitlab-ci.yml
Comment (atodorov) [inline .gitlab-ci.yml:112]: @achilleas-k I propose to replace the 8.4 runner with 8.5 instead, not add another one. Also once the runner PR lands I prefer to actually change all of the 8.4 runners (we can do that separately).

Comment (achilleas-k) [reply .gitlab-ci.yml:111]: Changed ostree tests to run in 8.5 and the rest in 8.4 until we support all image types in 8.5.

File: schutzbot/mockbuild.sh.orig
Comment (atodorov) [inline schutzbot/mockbuild.sh.orig:77]: Unresolved conflict.

Comment (achilleas-k) [reply schutzbot/mockbuild.sh.orig:77]: accidentally added a merge conflict `.orig` file during rebase

**Review by atodorov — APPROVED** [review]

**Review by teg — APPROVED** [review]

**Pull request merged**
